Handleiding Vignette: headers and Rmd formatting (#235)
inbo/LSVI: vignettes/Handleiding.Rmd

Verder bevat het package functies m.b.t. de berekening van de LSVI.  De hoofdfunctie `berekenLSVIbasis()` kan door iedereen gebruikt worden om de LSVI te berekenen op basis van eigen gegevens.  De gebruiker moet wel eerst de gegevens van de veldopnames omzetten naar het formaat dat beschreven is in de helpfunctie van de functie zelf.  Op termijn zullen er ook functies ontwikkeld worden op maat van een specifieke databank of project (met bruikbaarheid afhankelijk van de toegang tot de databank).

> Omwille van de complexiteit van de berekening roept `berekenLSVIbasis` een aantal hulpfuncties op die elk een deel van de controle of berekening uitvoeren.  Voor het gemak bij het ontwikkelen zullen deze voorlopig beschikbaar zijn; op termijn zullen omwille van de overzichtelijkheid enkel de functies beschikbaar gesteld blijven die een meerwaarde bieden voor de gebruiker (en relatief gemakkelijk toegepast kunnen worden).

Reviewer: I prefer to keep my adapted version where I refer to functions adding ().  (This also adds links in the website.)

```suggestion
Omwille van de complexiteit van de berekening roept `berekenLSVIbasis()` een aantal hulpfuncties op die elk een deel van de controle of berekening uitvoeren.  Voor het gemak bij het ontwikkelen zullen deze voorlopig beschikbaar zijn; op termijn zullen omwille van de overzichtelijkheid enkel de functies beschikbaar gesteld blijven die een meerwaarde bieden voor de gebruiker (en relatief gemakkelijk toegepast kunnen worden).
```

Author: Ik wou dit aanpassen en pushen maar heb op dit ogenblik blijkbaar geen schrijfrechten. Sorry!

Reviewer: Misschien heeft iemand je uit team BMK gegooid op Github?  Ik heb je intussen terug toegevoegd, en bij uitbreiding het hele team MNM.

write_csv2(soortenlijst4030, "C:/R/soortenlijst4030.csv")
```

We hebben hier het absolute path (`C:/R`) weergegeven, een andere mogelijkheid is om het path relatief ten opzichte van de 'working directory' weer te geven.

> Dit is de map waarin je aan 't werken bent.

Reviewer: ```suggestion
Dit is de map waarin je aan 't werken bent, bijvoorbeeld de hoofdfolder van je project of een subfolder waarin je je scripts en data opslaat.
```

> Je kan deze working directory zelf bepalen door het commando `setwd("C:/.../mijn_werkmap")`.

Reviewer: ```suggestion
Je kan deze working directory zelf instellen door het commando `setwd("C:/.../mijn_werkmap")` (wat je telkens moet herhalen na opstarten van R).
```

> Als je in een project werkt waarin je al je scripts opslaat (aan te raden), dan is de root van dit project automatisch je working directory.

Reviewer: ```suggestion
In RStudio kan je ook een 'R-project' aanmaken voor een nieuwe of bestaande folder (aan te raden als je meerdere scripts hebt), en na het openen van dit project is de root van dit project automatisch je working directory.
```

write_csv2(soortenlijst4030, "C:/R/soortenlijst4030.csv")
```

> We hebben hier het absolute path (`C:/R`) weergegeven, een andere mogelijkheid is om het path relatief ten opzichte van de 'working directory' weer te geven.  Dit is de map waarin je aan 't werken bent.  Je kan deze working directory zelf bepalen door het commando `setwd("C:/.../mijn_werkmap")`.
> Her gebruik van `setwd()` is echter niet meer aangeraden, een alternatief biedt de package "[`here`](https://here.r-lib.org)" met onder meer de functie `here::here()`.
> Als je in een project werkt waarin je al je scripts opslaat (aan te raden), dan is de root van dit project automatisch je working directory, en je kan met `here::here()` navigeren (bv. `here::here("..", "relative", "path", "to", "file.txt")`).

Reviewer: Goed dat je het aangeeft, maar dit lijkt me veel te moeilijk voor de doelgroep van dit stukje tekst (= gebruikers die nooit met R werken en die we hier zo eenvoudig mogelijk proberen uitleggen hoe ze een tabel kunnen opslaan om in Excel te openen, zodat ook zij de info uit de databank kunnen raadplegen en op hun manier gebruiken).

Ik twijfel een beetje wat we hiermee best doen: enkel uitleggen dat ze een project kunnen aanmaken en en relatieve paden kunnen gebruiken t.o.v. de root van het project (wat al een hele stap is voor iemand die niet vertrouwd is met programmeren!), of ook een concreet code-voorbeeld geven met package `here` (met dan installatie-instructies voor dat package).  Zoals het er nu staat (een link en trek je plan), denk ik niet dat dit gebruikt zal worden door iemand die enkel op zoek is naar een soortenlijstje om te downloaden.
Wat lijkt u het beste?

```suggestion
We hebben hier het absolute path (`C:/R`) weergegeven, een andere mogelijkheid is om het path relatief ten opzichte van de 'working directory' weer te geven.  Dit is de map waarin je aan 't werken bent.  Je kan deze working directory zelf bepalen door het commando `setwd("C:/.../mijn_werkmap")`.
Her gebruik van `setwd()` is echter niet meer aangeraden, een alternatief biedt de package "[`here`](https://here.r-lib.org)" met onder meer de functie `here::here()`.
Als je in een project werkt waarin je al je scripts opslaat (aan te raden), dan is de root van dit project automatisch je working directory, en je kan met `here::here()` navigeren (bv. `here::here("..", "relative", "path", "to", "file.txt")`).
```

Author: Voor mij is `setwd()` prima, hoor! Simple, understandable, does the job.
Ik heb dit van collega's @wlangera en @ThierryO ergens opgepikt en overgenomen, en misschien hebben zij er een sterke mening over. Enerzijds is het goed om nieuwelingen ineens correct op te leiden, maar dit is een algemeen punt die we toch naar onze smaak toepassen.
Ik wou een referentie zoeken en zie wel [deze hier](https://stackoverflow.com/a/13770896) en een beetje gebabbel op reddit. Er is een author bias zichtbaar. Ik ben er zelf niet mee.

- Het hoofdargument van de `never ever setwd!!!!1!` fangroep is reproduceerbaarheid: als de computeromgeving verandert, gaan misschien bestanden niet mee. Maar dat is een algemeen probleem; de oplossing is niet nog meer "compliance", maar documentatie. Als je een verwijs goed documenteerd, zie ik geen probleem, dan kan iedereen zijn script handmatig aanpassen (zie "settings" beneden). En als alle bestanden sowieso binnen één projectmap zitten, is bewegen binnen die projectmap geen enkel probleem.
- Ik zei "projectmap", want een project definieren met `Rproj` is ook niet universeel: bij mij (linux/emacs) werkt dit helemaal niet. Dit is specifiek voor Rstudio. Dikke "B-" voor reproduceerbaarheid.
- Er is de operating system, maar ja, iemand die een error door het verschil tussen `\\` en `/` niet kan googelen gaat ook met de gebruik van `here::here()` niet ineens vliegen.

Hoe zou ik dit intuitief oplossen? Door "settings" (in tekstbestanden, met een default, en een GUI of script om de gebruiker door die settings te begeleiden). Je laat de gebruiker zijn keuzes configureren, op een centrale plek (bv. met een [ConfigParser](https://cran.r-project.org/web/packages/ConfigParser/index.html)) en moet dan zeker gaan dat je package overal de projectmap vanuit die settings gebruikt, met OS-independent paths. Dit neemt heel veel van de potentiele fouten weg van de gebruiker en brengt ze naar de ontwikkelaar.
Maar zolang een package geen settings-structuur heeft, is `setwd()` ook niet het eind van de wereld.

Dit is maar mijn naïve brainstorm, jullie hebben meer ervaring - bedankt voor advies en discussie :)
Van mij zal ik dus `setwd()` terug de voorkeur geven; maar ik ben hier bij LSVI maar marginaal geïnvolveerd.

Reviewer: Bedankt voor de uitgebreide uitleg, zeer leerrijk.
Inderdaad, naar reproduceerbaarheid is het geen goed idee, maar dat geldt evenzeer voor het absolute path van mijn eerste optie.  Anderzijds moet je naar het doelpubliek kijken: reproduceerbaarheid is relevant voor diegenen die code schrijven en delen met anderen, en iemand die nog moet leren hoe een dataframe op te slaan om te openen in Excel, is meestal niet de persoon die code schrijft om te delen met teamgenoten of anderen.  Die zijn al zeer blij als ze het op hun computer aan de praat krijgen.  (En diegenen die de smaak te pakken krijgen en naar de coding club gaan, leren daar wel hoe het zou moeten.)

Qua gebruikersgroep die niet vertrouwd is met R, ga ik voor LSVI eigenlijk vooral uit van INBO-collega's, en dan ga ik ervan uit dat ze Windows hebben en de typische INBO-setup.  (Buitenstaanders zie ik al minder snel met R beginnen speciaal voor die toch wel zeer specifieke info, en Linux-gebruikers zijn meestal al iets meer gewend om zaken zelf uit te zoeken.)  Toegegeven, dit is wel een gemakkelijkheidsoplossing, maar in de praktijk heb ik nog maar enkel vragen gehad van INBO en ANB.  Moesten er vragen opduiken waaruit blijkt dat dit niet volstaat, dan pas ik de handleiding uiteraard aan in functie van de noden van de gebruikers.

Reviewer 2: Good practice is om je werk samen te zetten in een map of een mappenstructuur.

Binnen die structuur kies je een basismap. Als je RStudio gebruikt, dan is dat de logische plaats om je RStudio project te zetten. Als je git gebruikt is dat de basis van je repo.

Alle scripts binnen het project gebruiken enkel relatieve paden t.o.v. die basismap. Binnen die scripts heb je dan geen `setwd()` nodig.

Bij de start van de analyse zorg je ervoor dat de working directory overeenkomt met die basismap. RStudio doet dat automatisch. Als je dat manueel moet doen, dan kan je `setwd()` gebruiken.

Author: > Good practice is om je werk samen te zetten in een map of een mappenstructuur.

Ik ben er volledig mee, @ThierryO, en ik zou dit zelf ook zo doen, waar het kan! 

~~Maar~~ Akkord! Ik wou wel nog toevoegen: wat "good practice" is hangt af van de omstandigheden. We moeten wel ook in het oog houden dat de "working paradigm" die jij beschrijft heel R-specifiek is, en een eilandje. De meerderheid van de programma's waarmee mensen in contakt komen lossen dit anders op: door de gebruiker een menu te geven (bv. "File >> Project >> Save"). Die programma's werken **map-onafhankelijk** en **zonder "side-effects"**. Dit zou in R toch ook mogelijk moeten zijn.

Ik heb enkel de handleiding gelezen, en LSVI niet gebruikt. Maar als het over een software gaat die één gebruiker voor meerdere projectjes wil gebruiken (zoals je in [gimp 3.0](https://www.gimp.org/news/2025/03/16/gimp-3-0-released) beelden bewerkt, of in logseq verschillende grafieken hebt), dan zou het een optie zijn om dit qua design te ondersteunen. (niet nu; voor de toekomst).

Ik ga mijn zinnetje weer verwijderen en laat het aan jou @ElsLommelen om te beslissen.

Reviewer: Op zich heeft @ThierryO wel gelijk: ook mensen die niet met R werken (en die een klein beetje gestructureerd werken), zullen alles van eenzelfde project samenzetten in een map.  En dat lijkt me inderdaad ook wel een goed startpunt om die 'working directory' uit te leggen: de map waarin alles van je project bijeen zit.  Ik ga meteen hier een suggestie toevoegen.
